### ISSUE TITLE
Stylesheets should be removed when components/templates are disconnected

### ISSUE THREAD

**Issue by nolanlawson** [salesforce/lwc#2466]
### Description

When a component (shadow or light, native or synthetic) is disconnected from the DOM, we should clean up any stylesheets associated with that component. Currently we do this in some places but not others, and it's treated more as an implementation detail.

In particular, today we do not clean up `<style>`s that are added to the document `<head>` when they are no longer needed. This affects synthetic shadow DOM currently.

#### Steps to Reproduce

There are two ways to reproduce this:

1. Append a component to the DOM and then remove it.
2. In a component that renders multiple templates, render template A and then template B.

#### Expected Results

The stylesheets for the removed component/template should no longer be present in the DOM.

#### Actual Results

The `<style>` for the removed component/template is still in the DOM.

#### Other context

It's worth noting that [not every CSS-in-JS library or framework actually does this](https://twitter.com/sebmarkbage/status/1422278240866078725). So it might not be a big deal in practice.

#### Workarounds

Use scoped styles. [They don't have this same problem](https://github.com/salesforce/lwc/pull/2797#issuecomment-1116391759), at least with multi-template components.

**Comment by nolanlawson** [comment]
I suppose it's debatable whether we should fix case 1 as well as case 2. Case 2 is a bit more obviously correct (template B has replaced template A, so template A's styles should no longer be in the DOM). Whereas with case 2, it might actually be a performance hit to constantly remove and re-insert `<style>`s as elements are removed and re-added to the DOM.

E.g. imagine an SPA page transition where every `<x-button>` is removed from the DOM, but then the new page has its own `<x-button>`s. Do we remove the relevant `<style>` and then re-insert it? Seems like a waste.

**Comment by nolanlawson** [comment]
Interestingly, there is [some discussion among framework authors](https://twitter.com/sebmarkbage/status/1422278240866078725) indicating that not all frameworks clean up their CSS on unmount.

**Comment by nolanlawson** [comment]
We've decided not to implement this, for a few reasons:

1. There is a substantial amount of bookkeeping code required to implement this correctly. This is extra cost in terms of complexity as well as bytes sent over the wire.
2. There is a runtime performance cost as well, up to ~5% in some benchmarks (https://github.com/salesforce/lwc/pull/2797).
3. Using native shadow, styles are automatically removed when the shadow root is removed. No bookkeeping is required on the framework's part.
4. Many other frameworks and libraries do not bother to remove styles. (https://github.com/salesforce/lwc/pull/2797#issuecomment-1115192579)
5. In practice, LWC component authors can use [scoped styles](https://rfcs.lwc.dev/rfcs/lwc/0116-light-dom-scoped-styles) to cleanly swap templates in a multi-template component. (This works in both light DOM and shadow DOM, in both synthetic and native shadow.)

Closing this issue.
